roofit/roofitcore/src/RooMinimizer.cxx

```diff
@@ -298,6 +260,27 @@ RooFitResult* RooMinimizer::fit(const char* options)
 }
 
 
+bool RooMinimizer::fitFcn() const {
+   bool ret;
+
+   switch (_fcnMode) {
+   case FcnMode::classic: {
+      ret = _theFitter->FitFCN(*static_cast<RooMinimizerFcn *>(_fcn));
+      break;
+   }
+   case FcnMode::gradient: {
+      auto thing = dynamic_cast<RooGradMinimizerFcn *>(_fcn);
+      if (thing != nullptr) {
+         ret = _theFitter->FitFCN(*thing);
+      } else {
+         throw std::logic_error("In RooMinimizer::fitFcn: Minimizer fcnMode was set to gradient, but fit function type does not match RooGradMinimizerFcn!");
+      }
+      break;
+   }
+   }
+
+   return ret;
```

Comment:
```suggestion
   return _theFitter->FitFCN(*reinterpret_cast<ROOT::Math::IBaseFunctionMultiDim*>(_fcn));
```
Okay, maybe the `reinterpret_cast` is not so nice, but I think the fact that I even thought about it hints to some problem with the inheritance hierachy here.
I think the problem is that `RooAbsMinimizerFcn` doesn't inherit from `ROOT::Math::IBaseFunctionMultiDim`, but all implementations of the `RooAbsMinimizerFcn` need to. What keeps us from making `RooAbsMinimizerFcn` inherit from `ROOT::Math::IBaseFunctionMultiDim`? Then we can do a `static_cast` here.

Reply:
The point in `FitFCN` is that there are two overloads that we need to use here: one for `IBaseFunctionMultiDim` and one for `IGradientMultiDim`. `...Gradient...` inherits from `...BaseFunction...`, so if you just inherit all from `...BaseFunction...`, you end up in the same `FitFCN`, not in the one that uses the gradient.

Reply:
Ah yes, thanks for the clarification! I understand better now. So you still need the separate casts to `RooMinimizerFcn` and `RooGradMinimizerFcn` then. Then my suggestion here and and also in the other comment above doesn't work.

Still, I don't think it's necessary to do the dynamic cast here and the exception, because you already make sure that the function is either a  `RooMinimizerFcn` or a `RooGradMinimizerFcn`.

Reply:
Yes, agreed, I also commented that in another place where you suggested the same thing :) The dynamic_casts and corresponding if/else blocks can go.

Reply:
`reinterpret_cast` is probably dangerous (wrong) if you have multiple inheritance at work. It will just force the type to be something else, but it *will not adjust the pointer*, which is usually necessary when mult. inheritance is at work. You might end up with the correct pointer. Or not.

Don't do that. Never. (If you can avoid it).

When you know the inheritance graph, `static_cast<FinalType>(static_cast<TypeThatSelectsTheCorrectBranch>(...))` is the way to go. When you don't, there's only `dynamic_cast` if I'm not mistaken. If there's a way to avoid mult. inheritance, it's the preferred solution (exception: pure virtual base class), but if that's not possible, see above.